assets/js/hocs/test/with-product.js

```diff
@@ -0,0 +1,95 @@
+/**
+ * External dependencies
+ */
+import TestRenderer from 'react-test-renderer';
+
+/**
+ * Internal dependencies
+ */
+import withProduct from '../with-product';
+import * as mockUtils from '../../components/utils';
+
+// Mock the getProduct functions for tests.
+jest.mock( '../../components/utils', () => ( {
+	getProduct: jest.fn().mockImplementation(
+		() => Promise.resolve()
+	),
+} ) );
+
+const mockProduct = { name: 'T-Shirt' };
+const attributes = { productId: 1 };
+const TestComponent = withProduct( ( props ) => {
+	return <div
+		error={ props.error }
+		getProduct={ props.getProduct }
+		isLoading={ props.isLoading }
+		product={ props.product }
+	/>;
+} );
+const render = () => {
+	return TestRenderer.create(
+		<TestComponent
+			attributes={ attributes }
+		/>
+	);
+};
+
+describe( 'withProduct Component', () => {
+	afterEach( () => {
+		mockUtils.getProduct.mockClear();
+	} );
+
+	describe( 'lifecycle events', () => {
+		const renderer = render();
+
+		it( 'getProduct is called on mount with passed in product id', () => {
+			const { getProduct } = mockUtils;
+
+			expect( getProduct ).toHaveBeenCalledWith( attributes.productId );
+			expect( getProduct ).toHaveBeenCalledTimes( 1 );
+		} );
+
+		it( 'getProduct is hooked to the prop', () => {
+			const { getProduct } = mockUtils;
+			const props = renderer.root.findByType( 'div' ).props;
+
+			props.getProduct();
+
+			expect( getProduct ).toHaveBeenCalledTimes( 1 );
+		} );
+	} );
+
+	describe( 'when the API returns product data', () => {
+		mockUtils.getProduct.mockImplementation(
+			( productId ) => Promise.resolve( { ...mockProduct, id: productId } )
+		);
+		const renderer = render();
+
+		it( 'sets the product props', () => {
+			const props = renderer.root.findByType( 'div' ).props;
+
+			expect( props.error ).toEqual( null );
+			expect( typeof props.getProduct ).toEqual( 'function' );
+			expect( props.isLoading ).toEqual( false );
+			expect( props.product ).toEqual( { ...mockProduct, id: attributes.productId } );
+		} );
+
+		mockUtils.getProduct.mockReset();
```

Comment:
instead of a mockClear() here I wonder if it might be better to just do a `beforeEach` mockImplementation for getProduct (as the base which can be overridden on nested tests) and then an `afterEach` mockReset.

So basically where you currently have defined in the top-level describe block:

```js
afterEach( () => {
		mockUtils.getProduct.mockClear();
	} );
```

Change that to:

```js
beforeEach( () => {
    mockUtils.getProduct.mockImplementation( () => Promise.resolve() );
} );
afterEach( () => {
    mockUtils.getProduct.mockReset();
} );
```

Reply:
It looks much more organized. Done in 3ba87d15e0974051a31b7f139133fb1125941eb2.